common/src/chain/transaction/signature.rs

```diff
@@ -0,0 +1,512 @@
+    #[error("Invalid sighash value provided")]
+    InvalidSigHashValue(u8),
+    #[error("Invalid input index was provided (provided: `{0}` vs available: `{1}`")]
+    InvalidInputIndex(usize, usize),
+    #[error("Input corresponding to output number {0} does not exist (number of inputs is {1})")]
+    InvalidOutputIndexForModeSingle(usize, usize),
+    #[error("Decoding witness failed ")]
+    DecodingWitnessFailed,
+    #[error("Signature verification failed ")]
+    SignatureVerificationFailed,
+    #[error("Public key to address mismatch")]
+    PublicKeyToAddressMismatch,
+    #[error("Address authorization decoding failed")]
+    AddressAuthDecodingFailed(String),
+    #[error("Signature decoding failed")]
+    InvalidSignatureEncoding,
+    #[error("No signature!")]
+    SignatureNotFound,
+    #[error("Producing signature failed!")]
+    ProducingSignatureFailed(crypto::key::SignatureError),
+    #[error("Private key does not match with spender public key")]
+    SpendeePrivatePublicKeyMismatch,
+    #[error("Unsupported yet!")]
+    Unsupported,
+}
+
+pub fn signature_hash(
+    mode: sighashtype::SigHashType,
+    tx: &Transaction,
+    input_num: usize,
+) -> Result<H256, TransactionSigError> {
+    let mut stream = DefaultHashAlgoStream::new();
+
+    // TODO: even though this works fine, we need to make this function
+    // pull the inputs/outputs automatically through macros;
+    // the current way is not safe and may produce issues in the future
+
+    let target_input = tx
+        .get_inputs()
+        .get(input_num)
+        .ok_or_else(|| TransactionSigError::InvalidInputIndex(input_num, tx.get_inputs().len()))?;
+
+    hash_encoded_to(&mode.get(), &mut stream);
+    hash_encoded_to(&tx.version_byte(), &mut stream);
+    hash_encoded_to(&tx.get_flags(), &mut stream);
+
+    match mode.inputs_mode() {
+        sighashtype::InputsMode::CommitWhoPays => {
+            hash_encoded_to(&(tx.get_inputs().len() as u32), &mut stream);
+            for input in tx.get_inputs() {
+                hash_encoded_to(&input.get_outpoint(), &mut stream);
+            }
+        }
+        sighashtype::InputsMode::AnyoneCanPay => {
+            hash_encoded_to(&target_input.get_outpoint(), &mut stream);
+        }
+    }
+
+    match mode.outputs_mode() {
+        sighashtype::OutputsMode::All => {
+            hash_encoded_to(tx.get_outputs(), &mut stream);
+        }
+        sighashtype::OutputsMode::None => (),
+        sighashtype::OutputsMode::Single => {
+            let output = tx.get_outputs().get(input_num).ok_or_else(|| {
+                TransactionSigError::InvalidInputIndex(input_num, tx.get_outputs().len())
+            })?;
+            hash_encoded_to(&output, &mut stream);
+        }
+    }
+
+    hash_encoded_to(&tx.get_lock_time(), &mut stream);
+
+    // TODO: for P2SH add OP_CODESEPARATOR position
+    hash_encoded_to(&u32::MAX, &mut stream);
+
+    let result = stream.finalize().into();
+    Ok(result)
```

Comment:
I guess incorporating spent output amounts into sighash is going to be a follow-up?

Reply:
Isn't that part of outputs?

Reply:
It just occurs to me. If you mean including the amount of inputs, it turned out it's a bad idea and there's an attack related to that that nullifies its benefit. The most secure way is, unfortuntaely, just to download the input transactions and their merkle branch. So we'll probably just not add the amount of the inputs.

Reply:
If you're happy with this answer, please resolve this.

Reply:
OK, resolving. I'd be interested in learning more details about the attack.

Reply:
The attack is simply that attempting to use an output amount without verification of the output basically makes it possible for compromised software to trick the user with a hardware wallet into signing multiple outputs under the guise of "something went wrong, let's try again", given that the user can know nothing about the outputs being spent, which can result in burning a huge amount as a transaction fee.